File: android/src/main/java/com/jvanila/flutter/shopify/usecases/CompleteCheckoutByCardUseCase.java
package com.jvanila.flutter.shopify.usecases;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.jvanila.flutter.pluginarch.PluginContext;
import com.shopapp.gateway.ApiCallback;
import com.shopapp.gateway.entity.Address;
import com.shopapp.gateway.entity.Checkout;
import com.shopapp.gateway.entity.Error;
import com.shopapp.gateway.entity.Order;
import com.shopapp.shopify.api.ShopifyApi;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class CompleteCheckoutByCardUseCase extends ShopifyCallUseCase {

    private static final String ARG_ADDRESS_JSON = "addressJson";
    private static final String ARG_CHECKOUT_JSON = "checkoutJson";
    private static final String ARG_CREDIT_CARD_VALUE_TOKEN = "creditCardValueToken";
    private static final String ARG_EMAIL = "email";

    public CompleteCheckoutByCardUseCase(PluginContext<ShopifyApi> pluginContext) {
        super(pluginContext);
    }

    @Override
    protected void call(MethodCall input, final MethodChannel.Result result) {
        String token = input.argument(ARG_CREDIT_CARD_VALUE_TOKEN);
        String email = input.argument(ARG_EMAIL);
        String addressJson = input.argument(ARG_ADDRESS_JSON);
        String checkoutJson = input.argument(ARG_CHECKOUT_JSON);

        Address address = new Gson().fromJson(addressJson,
                new TypeToken<Address>() {
                }.getType());

        Checkout checkout = new Gson().fromJson(checkoutJson,
                new TypeToken<Checkout>() {
                }.getType());

        mPluginContext.api.instance.completeCheckoutByCard(checkout, email, address, token,
                new ApiCallback<Order>() {
                    @Override
                    public void onResult(Order order) {
                        if (order != null) {

                            String orderJsonString = new Gson().toJson(order,
                                    new TypeToken<Order>() {
                                    }.getType());

                            result.success(orderJsonString);
                        }
                    }

                    @Override
                    public void onFailure(Error error) {
                        System.out.println("onFailure -- " + error);
                    }
                });
    }
}

File: ios/Classes/UseCases/CompleteCheckoutByCardUseCase.swift
//
//  CompleteCheckoutByCardUseCase.swift
//  shopify
//
//  Created by Joseph jose on 17/08/19.
//

import UIKit
import ShopApp_Shopify
import ShopApp_Gateway
import MobileBuySDK
import Flutter

public class CompleteCheckoutByCardUseCase: UseCase {
    
    static let ARG_ADDRESS_JSON = "addressJson"
    static let ARG_CHECKOUT_JSON = "checkoutJson"
    static let ARG_CREDIT_CARD_VALUE_TOKEN = "creditCardValueToken"
    static let ARG_EMAIL = "email"
    
    override public init(_ context: PluginContext) {
        
        super.init(context)
    }

    override public func trigger(with methodCall: FlutterMethodCall, result: @escaping (Any?) -> Void) {
        
        if let args = methodCall.arguments as? [String:String] {
            let token = args[CompleteCheckoutByCardUseCase.ARG_CREDIT_CARD_VALUE_TOKEN]
            let email = args[CompleteCheckoutByCardUseCase.ARG_EMAIL]
            let addressJson = args[CompleteCheckoutByCardUseCase.ARG_ADDRESS_JSON]
            let checkoutJson = args[CompleteCheckoutByCardUseCase.ARG_CHECKOUT_JSON]
            
            
            let address = Address(addressJSON: addressJson)
            let checkout = Checkout(from: checkoutJson)
            
            let payAddress = PayAddress(from: address, email: email)
            (mContext?.api.instance as! ShopifyAPI).completeCheckout(checkout.payCheckout, billingAddress: payAddress, applePayToken: "", idempotencyToken: token!) { (order, error) in
                
                if error != nil || order == nil {
                    
                    if error != nil {
                        self.createError(withCase: "completeCheckoutUseCase", message: error!.errorMessage ?? "", error: error!, on: result)
                    }
                    else {
                        
                        result(false)
                    }
                }else if order != nil {
                    
                    let jsonString = order!.toJSONString()
                    result(jsonString)
                }
            }
        }
    }
}
